### PR TITLE
Output of --help/-h should go to os.Stdout rather than os.Stderr

### PR DESCRIPTION
**Pull request by chappjc**
As described in Issue https://github.com/decred/dcrd/issues/105#issuecomment-250304334, `flags.ErrHelp` means that **`err` from `Parse()` the contains auto-generated help message from btcsuite/go-flags**, so we can print it to stdout.

**Clarification:** Only print to stdout when help is explicitly requested, _not_ because of an error.  Exit with code 0 in this case since there is no error.

Just opinions, but... http://stackoverflow.com/q/2199624/2778484

### CODE REVIEW

**Comment by chappjc** [comment]
Updated for dcrctl, but made the decision to leave usage/help message printed to stderr in the case of an error parsing the config file.

https://github.com/chappjc/dcrd/blob/02efc40369adb46c03ae3019eee92fd140440f4e/cmd/dcrctl/config.go#L230

**Comment by chappjc** [comment]
Updated as per discussion in https://github.com/decred/dcrwallet/pull/344.
Only output to stdout when -h/--help is explicitly given, and exit code 0 in this case.

**Comment by jrick** [comment]
You can't call os.Exit there because it will immediately terminate the process, when other work still needs to be done (logs flushed in deferred functions, for example).  Return back to the caller (main) so it can exit with 0.

**Comment by chappjc** [comment]
@jrick I though about that (could do `return nil, nil, nil` and it will work out in the call stack) but there are a ton of other os.Exit calls in that function.  Are those all wrong to have?

**Comment by marcopeereboom** [comment]
Before we do a bunch of work I want more votes.

**Comment by dajohi** [comment]
I like stderr for errors, and stdout for --help

**Comment by jrick** [comment]
@chappjc oh right, this happens before the logger is initialized.  All writes are done with fmt, which automatically flushes every newline.  I think os.Exit is fine in this case then.

**Comment by chappjc** [comment]
[GNU coding standards says](https://www.gnu.org/prep/standards/html_node/_002d_002dhelp.html):

> The standard --help option should output brief documentation for how to invoke the program, on standard output, then exit successfully. Other options and arguments should be ignored once this is seen, and the program should not perform its normal function.

I acknowledge that Decred is not GNU, but an explicit request for usage info almost always sends the information to standard output.

**Comment by marcopeereboom** [comment]
GNU coding standards isn't a thing.  Despite what they want you to believe.

**Comment by chappjc** [comment]
@marcopeereboom First I had heard of it to be honest.  I'm just trying to find some justification one way or the other.

**Comment by marcopeereboom** [comment]
All joking aside, GNU is not UNIX is accurate.  UNIX uses stderr for all exceptional output; help means the program didn't run thus it goes to stderr.  GNU is silly all around; they changed things just to change it so taking it serious is hard.  As I said prior, the real reason for stdout is because there are no manual pages (or bad/incorrect ones) on GNU/Linux so they substitute that with a huge help built in the binary and typing 2>&1 is hard.

I can float either way but I can not help making fun of GNU "coding standards".  Hell, look at their style guide and tell me you aren't bleeding from every membrane in your body ;)

**Comment by chappjc** [comment]
Rebased and squashed.

**Comment by alexlyp** [comment]
OK after much heated debate we are going to make the change to os.Stdout for help.  @marcopeereboom's issues with the change have been noted.

**Comment by chappjc** [comment]
I'm happy to have distracted from real development. ;)

**Comment by alexlyp** [comment]
@chappjc oh after talking with @jrick needs a quick change.

        if err != nil {
                if e, ok := err.(*flags.Error); ok && e.Type != flags.ErrHelp {
                        fmt.Fprintln(os.Stderr, err)
                        os.Exit(0)
                } else if ok && e.Type == flags.ErrHelp {
                        fmt.Fprintln(os.Stdout, err)
                        os.Exit(0)
                }
        }


So basically if any errs that aren't ErrHelp come back from the parser they are still sent to stderr

**Comment by jrick** [comment]
To be clear, this is the behavior we want and the reason for the change:

* When -h or --help is passed, the program is expected to print help, it is not an exceptional message, and should go to stdout.

* When some kind of parsing or configuration error occurs and help is printed, this is an exceptional error and any help should be written to stderr.

**Comment by chappjc** [comment]
@alexlyp Makes sense.  I think I realized that for the dcrwallet PR.  But what about `!ok`?  See https://github.com/decred/dcrwallet/pull/344/files

**Comment by alexlyp** [comment]
well `!ok` means that it was unable to assert the flags.Error type so I think it should just pass through since err would be nil anyways.

**Comment by chappjc** [comment]
@alexlyp Or `Parse` has a bug, in which case it should probably exit.  Also, there might as well be no `ok` test if the assumption is that the type assert will never fail if `err != nil`.

**Comment by alexlyp** [comment]
Well i think we can trust upstream btcsuite/go-flags to be without bugs, but I agree we may want a safety check to drop out on !ok.  But this is fine for now. 

Can you make the similiar changes to cmd/dcrctl/config.go for this PR and the dcrwallet PR as well?

File: config.go [hunk 1]
```diff
@@ -388,9 +388,12 @@ func loadConfig() (*config, []string, error) {
 	preParser := newConfigParser(&preCfg, &serviceOpts, flags.HelpFlag)
 	_, err := preParser.Parse()
 	if err != nil {
-		if e, ok := err.(*flags.Error); ok && e.Type == flags.ErrHelp {
+		if e, ok := err.(*flags.Error); ok && e.Type != flags.ErrHelp {
 			fmt.Fprintln(os.Stderr, err)
-			return nil, nil, err
+			os.Exit(0)
```
Comment (chappjc) [inline config.go]: Shouldn't this be Exit(1)?

File: config.go [hunk 2]
```diff
@@ -388,9 +388,12 @@ func loadConfig() (*config, []string, error) {
 	preParser := newConfigParser(&preCfg, &serviceOpts, flags.HelpFlag)
 	_, err := preParser.Parse()
 	if err != nil {
-		if e, ok := err.(*flags.Error); ok && e.Type == flags.ErrHelp {
+		if e, ok := err.(*flags.Error); ok && e.Type != flags.ErrHelp {
```
Comment (chappjc) [inline config.go]: Handle `!ok` in the future, perhaps.

File: config.go (hunk 1 above)
Comment (alexlyp) [inline config.go]: yah that seems more correct,  when using os.Stderr which is outputing an actual err (and not help) it should be os.Exit(1)

**Comment by alexlyp** [comment]
OK
